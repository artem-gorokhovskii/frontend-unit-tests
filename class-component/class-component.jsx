import React from 'react';

export class Counter extends React.PureComponent {
    constructor(props) {
        super(props);
        this.state = { pressedTimes: 0 };
        this.increasePressedTimesBinded = this.increasePressedTimes.bind(this);
    }

    componentDidUpdate(prevProps) { // Только для теста метода жизненного цикла
        // после изменения props.name установим prevProps.name.length в state.pressedTimes
        if (prevProps.name !== this.props.name) {
            this.setState({ pressedTimes: prevProps.name.length });
        }
    }

    increasePressedTimes() {
        this.setState(prevState => ({ pressedTimes: prevState.pressedTimes + 1 }));
    }

    render() {
        return (
            <div className="counter">
                <div className='greetings'>Привет, {this.props.name}!</div>
                <div className='message'>Вы нажали {this.state.pressedTimes} раз</div>
                {this.state.pressedTimes >= 3 && <div className='notification'>Кликов больше 2!</div>}
                <button onClick={this.increasePressedTimesBinded}></button>
            </div>
        );
    }
}

