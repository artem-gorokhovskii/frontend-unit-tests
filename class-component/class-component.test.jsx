import React from 'react';

import { shallow } from 'enzyme';
import { Counter } from './class-component';

const MIN_PROPS = { name: 'Artem Gorokhovskii' };

describe('Class component Counter', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallow(<Counter {...MIN_PROPS} />);
    });
    it('Должен иметь корректный снапшот', () => {
        expect(wrapper.debug()).toMatchSnapshot();
    });
    it('Приветствие выводится корректно', () => {
        const text = wrapper.find('.greetings').text();
        expect(text).toBe(`Привет, ${MIN_PROPS.name}!`);
    });
    it('Нажатие кнопки увеличит счетчик и выведет корректное сообщение', () => {
        const prevMessage = wrapper.find('.message').text();

        const button = wrapper.find('button');
        button.simulate('click');
        
        const message = wrapper.find('.message').text();

        expect(prevMessage).toBe('Вы нажали 0 раз');
        expect(message).toBe('Вы нажали 1 раз');
    });
    it('Нажатие кнопки 3 раза выведет предупреждение', () => {
        const prevNotificationExists = !!wrapper.find('.notification').length;

        wrapper.setState({ pressedTimes: 3 });
        
        const notificationExists = wrapper.find('.notification').length;
        
        expect(prevNotificationExists).toBe(false);
        expect(notificationExists).toBe(1);
    });
    it('Метод increasePressedTimes корректно изменит state', () => {
        expect(wrapper.state('pressedTimes')).toBe(0);
        wrapper.instance().increasePressedTimes();
        expect(wrapper.state('pressedTimes')).toBe(1);
    });
    it('Life cycle method componentDidUpdate корректно изменит state', () => {
        expect(wrapper.state('pressedTimes')).toBe(0);
        wrapper.setProps({ name: 'Artem' })
        expect(wrapper.state('pressedTimes')).toBe(MIN_PROPS.name.length);
    });
});