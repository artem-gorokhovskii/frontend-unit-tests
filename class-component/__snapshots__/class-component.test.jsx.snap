// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Class component Counter Должен иметь корректный снапшот 1`] = `
"<div className=\\"counter\\">
  <div className=\\"greetings\\">
    Привет, 
    Artem Gorokhovskii
    !
  </div>
  <div className=\\"message\\">
    Вы нажали 
    0
     раз
  </div>
  <button onClick={[Function: bound increasePressedTimes]} />
</div>"
`;
